### PR TITLE
Task/public api for kubikles

### PR DESCRIPTION
**Pull request by FChataigner**
[sc-83317]

### CODE REVIEW

File: dataikuapi/dss/codestudio.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        """
+        Restart a code studio
+
+        :returns: a future to wait on the start
+        :rtype: :class:`dataikuapi.dss.future.DSSFuture`
+        """
+        ret = self.client._perform_json("POST", "/projects/%s/code-studios/%s/restart" % (self.project_key, self.code_studio_object_id))
+        if 'jobId' in ret:
+            return self.client.get_future(ret["jobId"])
+        else:
```
Comment (cstenac) [inline dataikuapi/dss/codestudio.py:95]: Should not return null if not job id, but an already completed future. Use DSSFuture.from_resp

File: dataikuapi/dss/codestudio.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        """
+        Stop a running code studio
+
+        :returns: a future to wait on the stop, or None if already stopped
+        :rtype: :class:`dataikuapi.dss.future.DSSFuture`
+        """
+        ret = self.client._perform_json("POST", "/projects/%s/code-studios/%s/stop" % (self.project_key, self.code_studio_object_id))
+        if 'jobId' in ret:
+            return self.client.get_future(ret["jobId"])
+        else:
```
Comment (cstenac) [inline dataikuapi/dss/codestudio.py:82]: Should not return null if not job id, but an already completed future. Use DSSFuture.from_resp

File: dataikuapi/dss/codestudio.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        self.client = client
+        self.project_key = project_key
+        self.code_studio_object_id = code_studio_object_id
+        self.settings = settings
+
+    def get_raw(self):
+        """
+        Gets all settings as a raw dictionary. This returns a reference to the raw settings, not a copy,
+        """
+        return self.settings
```
Comment (cstenac) [inline dataikuapi/dss/codestudio.py:113]: save ?

File: dataikuapi/dss/codestudio.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        """
+        Gets the status as a raw dictionary. This returns a reference to the raw status, not a copy,
+        """
+        return self.status
+
+    @property
+    def state(self):
+        return self.status["state"]
+    @property
+    def last_built(self):
```
Comment (cstenac) [inline dataikuapi/dss/codestudio.py:136]: "built" for a code studio ?

File: dataikuapi/dss/project.py [hunk 5]
```diff
@@ -1628,6 +1629,61 @@ def get_mlflow_extension(self):
         """
         return DSSMLflowExtension(client=self.client, project_key=self.project_key)
 
+
+
+    ########################################################
+    # Code studios
+    ########################################################
+    def list_code_studio_objects(self, as_type="listitems"):
```
Comment (cstenac) [inline dataikuapi/dss/project.py:1637]: Why this wording of "objects" ? Why not "list_code_studios" ?

File: dataikuapi/dss/project.py [hunk 6]
```diff
@@ -1628,6 +1629,61 @@ def get_mlflow_extension(self):
...
+        """
+        List the code studio objects in this project
+        
+        Returns:
+            the list of the code studio objects, each one as a JSON object
```
Comment (cstenac) [inline dataikuapi/dss/project.py:1642]: Update to modern docstring

File: dataikuapi/dss/project.py [hunk 7]
```diff
@@ -1628,6 +1629,61 @@ def get_mlflow_extension(self):
...
+            return [DSSCodeStudioObject(self.client, self.project_key, item["id"]) for item in items]
+        else:
+            raise ValueError("Unknown as_type") 
+
+    def get_code_studio_object(self, code_studio_object_id):
+        """
+        Get a handle to interact with a specific code studio object
+       
+        Args:
+            code_studio_object_id: the identifier of the desired code studio object
```
Comment (cstenac) [inline dataikuapi/dss/project.py:1658]: Update to modern docstring

File: dataikuapi/dssclient.py [hunk 8]
```diff
@@ -593,6 +593,39 @@ def create_cluster(self, cluster_name, cluster_type='manual', params=None):
...
+
+    ########################################################
+    # Code studio templates
+    ########################################################
+
+    def list_code_studio_templates(self, as_type='listitems'):
+        """
+        List all code studio templates on the DSS instance
+
+        Returns:
```
Comment (cstenac) [inline dataikuapi/dssclient.py:605]: Modern docstring

File: dataikuapi/dssclient.py [hunk 9]
```diff
@@ -593,6 +593,39 @@ def create_cluster(self, cluster_name, cluster_type='manual', params=None):
...
+        elif as_type == "objects" or as_type == "object":
+            return [DSSCodeStudioTemplate(self, item["id"]) for item in items]
+        else:
+            raise ValueError("Unknown as_type") 
+
+    def get_code_studio_template(self, template_id):
+        """
+        Get a handle to interact with a specific code studio template
+        
+        Args:
```
Comment (cstenac) [inline dataikuapi/dssclient.py:620]: Modern docstring

File: dataikuapi/dss/codestudio.py (hunk 3 above)
Comment (FChataigner) [reply dataikuapi/dss/codestudio.py:113]: nothing to actually change at the moment, this call merely returns the code studio definition, so you can see the templateId and the libName. There is a `params` with a `runAs` field that's not implemented / used at the moment, but maybe one day...
So today the object is more a "immutable definition" than a "settings", but is named settings for future-proof-ness

File: dataikuapi/dss/codestudio.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    """
+    A handle to manage a code studio object of a project
+    """
+    def __init__(self, client, project_key, code_studio_object_id):
+        """Do not call directly, use :meth:`dataikuapi.dss.project.DSSProject.get_code_studio_object`"""
+        self.client = client
+        self.project_key = project_key
+        self.code_studio_object_id = code_studio_object_id
+
+    def delete(self):
```
[the remark this answers is not in the document]
Comment (FChataigner) [reply dataikuapi/dss/codestudio.py:46]: yes, the backend stops the code studio if running, and deletes the webapps using it if any

**Pull request merged**